File: 1371-minimum-remove-to-make-valid-parentheses/1371-minimum-remove-to-make-valid-parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
         Stack<Integer> st = new Stack<>();
        StringBuilder res = new StringBuilder(s);

        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            if(!st.isEmpty() && s.charAt(st.peek()) == '(' && c == ')'){
                st.pop();
            } else if(c == '(' || c == ')') {
                st.push(i);
            }
        }

        while(!st.isEmpty()){
            res.setCharAt(st.pop(), '#');
        }

        StringBuilder ans = new StringBuilder();
        for(int i = 0; i < res.length(); i++){
            if(res.charAt(i) != '#'){
                ans.append(res.charAt(i));
            }
        }

        return ans.toString();
    }
}

File: 1371-minimum-remove-to-make-valid-parentheses/1371-minimum-remove-to-make-valid-parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        cout.tie(NULL);
        
        stack<pair<char,int>> st;
        //string ans = s;
        string res = "";

        for(int i = 0; i < s.size(); i++){
            char ch = s[i];
            if(!st.empty() && st.top().first == '(' && ch == ')'){
                st.pop();
            }
            else if(ch == '(' || ch == ')'){
                st.push({ch,i});
            }
        }

         // Remove the invalid parentheses
        while (!st.empty()) {
            s[st.top().second] = '#';
            st.pop();
        }

        // making the resultant string by excluding '#'
        for (int i = 0; i < s.length(); ++i)
        {
            if (s[i] != '#')
                res.push_back(s[i]);
        }
        return res;
    }
};